modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java

```diff
@@ -172,21 +210,50 @@ private Object process0(long cntr, CacheContinuousQueryEntry entry, boolean back
                 if (res == RETRY)
                     continue;
             }
-            else
+            else {
+                if (batch.endCntr < ackedUpdCntr)
+                    batch.tryRollOver(entry.topologyVersion());
+
+                if (pendingSize.get() > MAX_PENDING_BUFF_SIZE) {
+                    LT.warn(log, "Buffer for pending events reached max of its size " +
+                        "[cacheId=" + entry.cacheId() + ", maxSize=" + MAX_PENDING_BUFF_SIZE +
+                        ", partId=" + entry.partition() + ']');
+
+                    // Remove first BUFF_SIZE keys.
+                    int keysToRemove = BUF_SIZE;
+
+                    Iterator<Map.Entry<Long, CacheContinuousQueryEntry>> iter = pending.entrySet().iterator();
+
+                    res = new ArrayList<>();
+
+                    while (iter.hasNext() && keysToRemove > 0) {
+                        // Collecting results ignore backup flag due to batch may not be switched.
+                        res = addResult(res, iter.next().getValue(), false);
+
+                        iter.remove();
+                        pendingSize.decrementAndGet();
+
+                        keysToRemove--;
+                    }
+                }
+
+                pendingSize.incrementAndGet();
                 pending.put(cntr, entry);
+            }
 
             break;
         }
 
         Batch batch0 = curBatch.get();
 
+        // Batch has been changed on entry processing to the new one.
         if (batch0 != batch) {
```

Comment:
Can be simplified with only one `while`, without `if`

Reply:
I'll change this under the IGNITE-13126.